tests/UnifiedSpecTests/UnifiedSpecTest.php

```diff
@@ -0,0 +1,346 @@
+
+    /** @var MongoDB\Client */
+    private static $internalClient;
+
+    /** @var FailPointObserver */
+    private $failPointObserver;
+
+    private static function doSetUpBeforeClass()
+    {
+        parent::setUpBeforeClass();
+
+        /* Provide internal client unmodified URI, since it may need to execute
+         * commands on multiple mongoses (e.g. killAllSessions) */
+        self::$internalClient = new Client(static::getUri(true));
+        self::killAllSessions();
+    }
+
+    private function doSetUp()
+    {
+        parent::setUp();
+
+        /* TODO: The transactions spec advises calling killAllSessions only at
+         * the start of the test suite and after failed tests; however, the
+         * "unpin after transient error within a transaction" pinning test
+         * causes the subsequent transaction test to block. This can be removed
+         * once that is addressed. */
+        self::killAllSessions();
+
+        $this->failPointObserver = new FailPointObserver();
+        $this->failPointObserver->start();
+    }
+
+    private function doTearDown()
+    {
+        if ($this->hasFailed()) {
+            self::killAllSessions();
+        }
+
+        $this->failPointObserver->stop();
+        $this->failPointObserver->disableFailPoints();
+
+        /* TODO: Consider manually invoking gc_collect_cycles since each test is
+         * prone to create cycles (perhaps due to EntityMap), which can leak and
+         * prevent sessions from being released back into the pool. */
+
+        parent::tearDown();
+    }
+
+    /**
+     * @dataProvider providePassingTests
+     */
+    public function testPassingTests(stdClass $test, string $schemaVersion, array $runOnRequirements = null, array $createEntities = null, array $initialData = null)
+    {
+        if (! $this->isSchemaVersionSupported($schemaVersion)) {
+            $this->markTestIncomplete(sprintf('Test format schema version "%s" is not supported', $schemaVersion));
+        }
+
+        if (isset($runOnRequirements)) {
+            $this->checkRunOnRequirements($runOnRequirements);
+        }
+
+        if (isset($test->skipReason)) {
+            $this->assertInternalType('string', $test->skipReason);
+            $this->markTestSkipped($test->skipReason);
+        }
+
+        if (isset($test->runOnRequirements)) {
+            $this->assertInternalType('array', $test->runOnRequirements);
+            $this->checkRunOnRequirements($test->runOnRequirements);
+        }
+
+        if (isset($initialData)) {
+            $this->prepareInitialData($initialData);
+        }
+
+        // Give Context unmodified URI so it can enforce useMultipleMongoses
+        $context = new Context(self::$internalClient, static::getUri(true));
+
+        if (isset($createEntities)) {
+            $context->createEntities($createEntities);
+        }
+
+        // TODO handle distinct commands in sharded transactions
```

Comment:
Basically copy in https://github.com/mongodb/mongo-php-library/blob/1.7.1/tests/SpecTests/TransactionsSpecTest.php#L318. Rather than run that for _each_ test, I think we can be a bit smarter and have the function quickly check if a sharded environment is being used _and_ scan the `operations` array for both `startTransaction` and `distinct` (order doesn't really matter, since it's just making an educated guess).